src/main/java/htsjdk/variant/variantcontext/ByteArrayAllele.java

```diff
@@ -0,0 +1,305 @@
+    }
+
+    /**
+     * @return hash code
+     */
+    @Override
+    public int hashCode() {
+        int hash = 1;
+        for (int i = 0; i < bases.length; i++)
+            hash += (i+1) * bases[i];
+        return hash;
+    }
+
+
+    /**
+     * Returns true if this and other are equal.  If ignoreRefState is true, then doesn't require both alleles has the
+     * same ref tag
+     *
+     * @param other            allele to compare to
+     * @param ignoreRefState   if true, ignore ref state in comparison
+     * @return true if this and other are equal
+     */
+    @Override
+    public boolean equals(final Allele other, final boolean ignoreRefState) {
+        return this == other || (isRef == other.isReference() || ignoreRefState) && isNoCall == other.isNoCall() && (bases == other.getBases() || Arrays.equals(bases, other.getBases()));
+    }
+
+    /**
+     * @param test  bases to test against
+     *
+     * @return  true if this Allele contains the same bases as test, regardless of its reference status; handles Null and NO_CALL alleles
+     */
+    @Override
+    public boolean basesMatch(final byte[] test) { return !isSymbolic && (bases == test || Arrays.equals(bases, test)); }
+
+    /**
+     * @param test  bases to test against
+     *
+     * @return  true if this Allele contains the same bases as test, regardless of its reference status; handles Null and NO_CALL alleles
+     */
+    @Override
+    public boolean basesMatch(final String test) { return basesMatch(test.toUpperCase().getBytes()); }
+
+
+    /**
+     * @param test  allele to test against
+     *
+     * @return  true if this Allele contains the same bases as test, regardless of its reference status; handles Null and NO_CALL alleles
+     */
+    @Override
+    public boolean basesMatch(final Allele test) { return basesMatch(test.getBases()); }
+
+    /**
+     * @return the length of this allele.  Null and NO_CALL alleles have 0 length.
+     */
+    @Override
+    public int length() {
+        return isSymbolic ? 0 : bases.length;
+    }
+
+    // ---------------------------------------------------------------------------------------------------------
+    //
+    // useful static functions
+    //
+    // ---------------------------------------------------------------------------------------------------------
+
+    @Override
+    public int compareTo(final Allele other) {
+        if ( isReference() && other.isNonReference() )
+            return -1;
+        else if ( isNonReference() && other.isReference() ) 
+            return 1;
+        else if (other instanceof ByteArrayAllele) {
+            final ByteArrayAllele baOther = (ByteArrayAllele) other;
+            return compareBases(baOther.bases);
+        } else {
+            return compareBases(other.getBases());
+        }
+    }
+
+    private int compareBases(final byte[] otherBases) {
+        final int stop = otherBases.length < bases.length ? otherBases.length : bases.length;
```

Comment:
Should this have a fast return if bases = otherBases?

Reply:
can't this stop calculation be replaced with Math.min?

Reply:
yeah... I kinda imagin that it has an added cost but perhaps it is inlined.

Reply:
I would hope it gets inline and has no added cost, but you can never be sure... Math.min is annotated as a hotspot intrinsic candidate which makes me suspect it will probably be very fast.  I think in any case it's clearer though.